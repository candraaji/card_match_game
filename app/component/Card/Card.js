import React, {useState, useRef, useEffect} from 'react';
import {Pressable, Animated, Text, Image, StyleSheet} from 'react-native';

const Card = ({dataCard, onPressCard, countOpenCard, setStepToOpen}) => {
  const [isFront, setIsFront] = useState(false);

  const flipAnimation = useRef(new Animated.Value(0)).current;

  let flipRotation = 0;
  flipAnimation.addListener(({value}) => (flipRotation = value));

  const flipToBackStyle = {
    transform: [
      {
        rotateY: flipAnimation.interpolate({
          inputRange: [0, 180],
          outputRange: ['180deg', '360deg'],
        }),
      },
    ],
  };

  const flipToFront = () => {
    setIsFront(true);
    setStepToOpen(x => x + 1);
    Animated.timing(flipAnimation, {
      toValue: 180,
      duration: 300,
      useNativeDriver: true,
    }).start();
  };
  const flipToBack = () => {
    setIsFront(false);
    Animated.timing(flipAnimation, {
      toValue: 0,
      duration: 300,
      useNativeDriver: true,
    }).start();
  };

  const onPressCardData = () => {
    if (countOpenCard <= 2) {
      flipToFront();
    }
    onPressCard(dataCard);
  };

  useEffect(() => {
    let timer = 1200;
    let cardFlipBackTimer = setTimeout(() => {
      flipToBack();
    }, timer);
    return () => {
      clearTimeout(cardFlipBackTimer);
    };
  }, [onPressCard]);

  if (dataCard.position === 'show') {
    return (
      <>
        <Animated.View style={{...styles.cardFront}}>
          <Text style={styles.dataCardStyle}>{dataCard?.number}</Text>
        </Animated.View>
      </>
    );
  } else {
    return (
      <Pressable
        onPress={() =>
          dataCard?.status === 'editable' ? onPressCardData(dataCard) : null
        }>
        {isFront ? (
          <Animated.View style={{...styles.cardFront, ...flipToBackStyle}}>
            <Text style={styles.dataCardStyle}>{dataCard?.number}</Text>
          </Animated.View>
        ) : (
          <Animated.View style={{...styles.cardFront, ...flipToBackStyle}}>
            <Image
              source={require('../../assets/cardBackground.png')}
              resizeMode={'cover'}
              style={styles.cardBack}
            />
          </Animated.View>
        )}
      </Pressable>
    );
  }
};

const styles = StyleSheet.create({
  cardFront: {
    alignSelf: 'center',
    borderWidth: 1,
    padding: 10,
    backgroundColor: 'white',
    justifyContent: 'center',
    height: 175,
    width: 125,
    borderRadius: 5,
  },
  cardBack: {
    alignSelf: 'center',
    borderWidth: 0.5,
    justifyContent: 'center',
    height: 175,
    width: 125,
    borderRadius: 5,
  },
  dataCardStyle: {
    textAlign: 'center',
  },
});

export default Card;
