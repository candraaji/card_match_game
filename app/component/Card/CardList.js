import React from 'react';
import Card from './Card';

const CardList = React.memo(function CardList({
  onPressCard,
  listCard,
  countOpenCard,
  setStepToOpen,
}) {
  return listCard.map((dataCard, index) => (
    <Card
      dataCard={dataCard}
      setStepToOpen={setStepToOpen}
      countOpenCard={countOpenCard}
      key={index}
      onPressCard={onPressCard}
    />
  ));
});

export default CardList;
