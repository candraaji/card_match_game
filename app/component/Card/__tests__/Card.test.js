import React from 'react';
import renderer from 'react-test-renderer';
import {cleanup, fireEvent, render, act} from '@testing-library/react-native';
import Card from '../Card';

afterEach(cleanup);

describe('Card Component', () => {
  const dataCard = {index: 1, number: 10, status: 'editable', position: 'hide'};
  const setStepToOpen = jest.fn();
  const countOpenCard = 1;
  const index = 1;
  const onPressCard = jest.fn();
  jest.useFakeTimers();

  test('it should render card compoennt correctly', () => {
    jest.spyOn(React, 'useRef').mockReturnValueOnce({current: 0});
    const wrapper = renderer.create(
      <Card
        dataCard={dataCard}
        setStepToOpen={setStepToOpen}
        countOpenCard={countOpenCard}
        key={index}
        onPressCard={onPressCard}
      />,
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('should can press the card', () => {
    const testIdName = 'PressableCard';

    const {getByTestId} = render(
      <Card
        dataCard={dataCard}
        setStepToOpen={setStepToOpen}
        countOpenCard={countOpenCard}
        key={index}
        onPressCard={onPressCard}
      />,
    );

    const foundButton = getByTestId(testIdName);

    fireEvent.press(foundButton);
    act(() => {
      jest.runAllTimers();
    });

    expect(foundButton).toBeTruthy();
  });
});
