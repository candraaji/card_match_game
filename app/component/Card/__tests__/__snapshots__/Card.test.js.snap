// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component it should render card compoennt correctly 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  testID="PressableCard"
>
  <View
    collapsable={false}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "white",
        "borderRadius": 5,
        "borderWidth": 1,
        "height": 175,
        "justifyContent": "center",
        "padding": 10,
        "transform": Array [
          Object {
            "rotateY": "180deg",
          },
        ],
        "width": 125,
      }
    }
  >
    <Image
      resizeMode="cover"
      source={
        Object {
          "testUri": "../../../app/assets/cardBackground.png",
        }
      }
      style={
        Object {
          "alignSelf": "center",
          "borderRadius": 5,
          "borderWidth": 0.5,
          "height": 175,
          "justifyContent": "center",
          "width": 125,
        }
      }
    />
  </View>
</View>
`;
